discojs/src/training/trainer.ts

```diff
@@ -22,6 +22,20 @@ export interface RoundLogs {
   participants: number;
 }
 
+/** Key: layer index, Value: array of weight update norms */
+export type WeightHistory = Map<number, number[]>;
```

Comment:
I'm striving to use [immutable's datastructures](https://immutable-js.com/) to avoid unintended changes of data-only structures. that does require a bit of changes as now everytime one add an element to a datastructure, it will create a new collection rather than updating to original one.

while reading a bit more on it, it seems that we don't really need the keys, it is akin to a list position so I guess we can even drop the `Map` and use an array/`List` directly.

also, it isn't really a "history of weights" but a "history of norms"

```suggestion
export type WeightHistory = List<List<number>>;
```